pbrd/pbr_vty.c

```diff
@@ -153,325 +128,949 @@ DEFUN_NOSH(no_pbr_map, no_pbr_map_cmd, "no pbr-map PBRMAP [seq (1-700)]",
+
+check:
 	pbr_map_check(pbrms, true);
+	return CMD_SUCCESS;
+}
 
+/* clang-format off */
+DEFPY  (pbr_map_match_dst_port,
+	pbr_map_match_dst_port_cmd,
+	"[no] match dst-port ![(1-65535)$port]",
+	NO_STR
+	"Match the rest of the command\n"
+	"Choose the destination port to use\n"
+	"The Destination Port\n")
+{
+	/* clang-format on */
+	struct pbr_map_sequence *pbrms = VTY_GET_CONTEXT(pbr_map_sequence);
+
+	if (!pbrms)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (no) {
+		if (!CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_DST_PORT))
+			return CMD_SUCCESS;
+		UNSET_FLAG(pbrms->filter_bm, PBR_FILTER_DST_PORT);
+		goto check;
+	}
+
+	if (CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_DST_PORT) &&
+	    (pbrms->dst_prt == port)) {
+		return CMD_SUCCESS;
+	}
+	pbrms->dst_prt = port;
+	SET_FLAG(pbrms->filter_bm, PBR_FILTER_DST_PORT);
+
+check:
+	pbr_map_check(pbrms, true);
 	return CMD_SUCCESS;
 }
 
-DEFPY(pbr_map_match_dst, pbr_map_match_dst_cmd,
-	"[no] match dst-ip <A.B.C.D/M|X:X::X:X/M>$prefix",
+/* clang-format off */
+DEFPY  (pbr_map_match_dscp,
+	pbr_map_match_dscp_cmd,
+	"[no] match dscp ![DSCP$dscp]",
 	NO_STR
 	"Match the rest of the command\n"
-	"Choose the dst ip or ipv6 prefix to use\n"
+	"Match based on IP DSCP field\n"
+	"DSCP value (below 64) or standard codepoint name\n")
+{
+	/* clang-format on */
+	struct pbr_map_sequence *pbrms = VTY_GET_CONTEXT(pbr_map_sequence);
+
+	if (!pbrms)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (no) {
+		if (!CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_DSCP))
+			return CMD_SUCCESS;
+		UNSET_FLAG(pbrms->filter_bm, PBR_FILTER_DSCP);
+		pbrms->dsfield &= ~PBR_DSFIELD_DSCP;
+		goto check;
+	}
+
+	long lDscp;
+	char *pEnd;
+	uint8_t rawDscp;
+
+	assert(dscp);
+	lDscp = strtol(dscp, &pEnd, 0);
+	if (*pEnd)
+		rawDscp = pbr_map_decode_dscp_enum(dscp);
+	else
+		rawDscp = lDscp << 2;
+	if (rawDscp > PBR_DSFIELD_DSCP) {
+		vty_out(vty, "Invalid dscp value: %s%s\n", dscp,
+			(pEnd ? "" : " (numeric value must be in range 0-63)"));
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	if (CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_DSCP) &&
+	    (((pbrms->dsfield & PBR_DSFIELD_DSCP) >> 2) == rawDscp)) {
+		return CMD_SUCCESS;
+	}
+
+	/* Set the DSCP bits of the DSField */
+	pbrms->dsfield = (pbrms->dsfield & ~PBR_DSFIELD_DSCP) | (rawDscp << 2);
+	SET_FLAG(pbrms->filter_bm, PBR_FILTER_DSCP);
+
+check:
+	pbr_map_check(pbrms, true);
+	return CMD_SUCCESS;
+}
+
+/* clang-format off */
+DEFPY  (pbr_map_match_ecn,
+	pbr_map_match_ecn_cmd,
+	"[no] match ecn ![(0-3)$ecn]",
+	NO_STR
+	"Match the rest of the command\n"
+	"Match based on IP ECN field\n"
+	"Explicit Congestion Notification\n")
+{
+	/* clang-format on */
+	struct pbr_map_sequence *pbrms = VTY_GET_CONTEXT(pbr_map_sequence);
+
+	if (!pbrms)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (no) {
+		if (!CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_ECN))
+			return CMD_SUCCESS;
+		UNSET_FLAG(pbrms->filter_bm, PBR_FILTER_ECN);
+		pbrms->dsfield &= ~PBR_DSFIELD_ECN;
```

Comment:
UNSET_FLAG?

Reply:
lib/zebra.h:

```
/* Flag manipulation macros. */
#define CHECK_FLAG(V,F)      ((V) & (F))
#define SET_FLAG(V,F)        (V) |= (F)
#define UNSET_FLAG(V,F)      (V) &= ~(F)
#define RESET_FLAG(V)        (V) = 0
#define COND_FLAG(V, F, C)   ((C) ? (SET_FLAG(V, F)) : (UNSET_FLAG(V, F)))
```

Reply:
I meant changing `pbrms->dsfield &= ~PBR_DSFIELD_ECN` to reuse UNSET_FLAG also :)

Reply:
lib/pbr.h:

```
#define PBR_DSFIELD_DSCP (0xfc) /* Upper 6 bits of DS field: DSCP */
#define PBR_DSFIELD_ECN (0x03)  /* Lower 2 bits of DS field: BCN */
```
Not exactly flags - it's a subfield. Are we (by convention) using the *_FLAG macros for general masking?